pymc3/distributions/multivariate.py

```diff
@@ -1091,3 +1091,219 @@ def logp(self, x):
+        Cholesky decomposition of right covariance matrix. Defines variance
+        within rows. Exactly one of rcov or rchol is needed.
+    lcov : qxq array
+        Left (or column) covariance matrix. Defines variance within
+        columns. Exactly one of lcov or lchol is needed.
+    lchol : qxq array
+        Cholesky decomposition of left covariance matrix. Defines variance
+        within columns. Exactly one of lcov or lchol is needed.
+
+    Examples
+    --------
+    Define a matrixvariate normal variable for given left and right covariance
+    matrices::
+
+        rcov = np.array([[1., 0.5], [0.5, 2]])
+        lcov = np.array([[1, 0, 0], [0, 4, 0], [0, 0, 16]])
+        q = lcov.shape[0]
+        p = rcov.shape[0]
+        mu = np.zeros((q, p))
+        vals = pm.MatrixNormal('vals', mu=mu, rcov=rcov,
+                               lcov=lcov, shape=(q, p))
+
+    Above, the ith row in vals has a variance that is scaled by 4^i.
+    Alternatively, left or right cholesky matrices could be substituted for
+    either covariance matrix. The MatrixNormal is quicker way compute
+    MvNormal(mu, np.kron(lcov, rcov)) that takes advantage of kronecker product
+    properties for inversion. For example, if draws from MvNormal had the same
+    covariance structure, but were scaled by different powers of an unknown
+    constant, both the covariance and scaling could be learned as follows
+    (see the docstring of `LKJCholeskyCov` for more information about this)::
+
+        # Setup data
+        true_rcov = np.array([[1.0, 0.5, 0.1],
+                              [0.5, 1.0, 0.2],
+                              [0.1, 0.2, 1.0]])
+        p = true_rcov.shape[0]
+        q = 3
+        true_scale = 3
+        true_lcov = np.diag([true_scale**(2*n) for n in range(q)])
+        mu = np.zeros((q, p))
+        true_kron = np.kron(true_lcov, true_rcov)
+        data = np.random.multivariate_normal(mu.flatten(), true_kron)
+        data = data.reshape(q, p)
+
+        # Setup right cholesky matrix
+        sd_dist = pm.HalfCauchy.dist(beta=2.5, shape=3)
+        rchol_packed = pm.LKJCholeskyCov('rcholpacked', n=3, eta=2,
+                                         sd_dist=sd_dist)
+        rchol = pm.expand_packed_triangular(3, rchol_packed)
+
+        # Setup left covariance matrix
+        scale = pm.Lognormal('scale', mu=np.log(true_scale), sd=0.5)
+        lcov = tt.nlinalg.diag([scale**(2*n) for n in range(q)])
+
+        vals = pm.MatrixNormal('vals', mu=mu, rchol=rchol, lcov=lcov,
+                               observed=data, shape=(q, p))
+    """
+
+    def __init__(self, mu=0, rcov=None, rchol=None, rtau=None,
+                 lcov=None, lchol=None, ltau=None, *args, shape=None,
+                 **kwargs):
+
+        self._setup_matrices(rcov, rchol, rtau, lcov, lchol, ltau)
+
+        if shape is None:
+            raise TypeError('shape is a required argument')
+        assert len(shape) == 2, "shape must have length 2: qxp"
```

Comment:
I think it would be better to raise a `ValueError` here instead of an `AssertionError`

Reply:
Also, if there are multiple observations, should the shape be $q \times p \times n_{obs}$?  If

Reply:
good points. If we allow for multiple observations, I do not know how to handle that in `random` and `_trquaddist`, where `matmul` and `solve_lower`, etc., will have to know that matrices are stacked somehow.

Reply:
you're right.  on second thought, I could certainly be wrong, but the multiple observations thing is probably pretty rare.  Could punt on that one until someone opens an issue (maybe never?)  What do you think about raising a NotImplementedError if the length of `shape` is greater than 2 and saying something informative?

Reply:
I had hoped that the multiple observations case would be rare as well, so I'm okay leaving that alone for now. A NotImplementedError could be more helpful in the future so I'll try that in the next commit. Would it be feasible when using Theano to implement something like `_process_parameters` [here](https://github.com/scipy/scipy/blob/v0.18.1/scipy/stats/_multivariate.py#L785) that infers the qxp shape rather than requiring the shape argument? Then we could maybe allow for simplified entries for the various covariances, e.g., a 1d array could become the diagonal.

Reply:
Since there's one observation, I think that should work... but I'm not positive.  Maybe @aseyboldt could weigh in here?

Reply:
Not sure how exactly that would work. My guess it that you can't do this without quite significant changes in a couple of places. At the moment we assume that the shape of all free variables is statically known, and this might break assumption.